New module = elb_target_group

wimnat opened the pull request:
##### ISSUE TYPE
- New Module Pull Request
##### COMPONENT NAME

elb_target_group
##### ANSIBLE VERSION

2.3
##### SUMMARY

A new module for AWS elastic load balancer target groups.  Goes well with my other PR for ec2_application_lb.

Unfinished but eager to get some feedback from the community.

Commit:
New module = elb_target_group

naslanidis committed:
Update elb_target_group.py

Adds support for adding and removing targets from a target group

wimnat committed:
Better error handling

wimnat committed:
Bump version_added

wimnat committed:
Minor fixes

wimnat commented:
Migrated from https://github.com/ansible/ansible-modules-extras/pull/2947

linuxdynasty commented:
Only suggestion I have, is to use the AWSRetry.backoff() decorator for all of the boto3 connections.

ryansb commented on lib/ansible/modules/cloud/amazon/elb_target_group.py:
```diff
@@ -0,0 +1,507 @@
+      - The number of consecutive health check failures required before considering a target unhealthy.
+    required: false
+    default: 2
+  successful_response_codes:
```
Is this a list, or an integer? Can multiple codes be passed here?

wimnat replied:
I've added more detail in the description now.

ryansb commented on lib/ansible/modules/cloud/amazon/elb_target_group.py:
```diff
@@ -0,0 +1,507 @@
+    port: 80
+    vpc_id: vpc-01234567
+    health_check_path: /
+    successful_response_codes: "200, 250-260"
```
This definitely bears explaining more in the docs, since it's an odd format and hard for us to check clientside.

wimnat replied:
As above

ryansb commented:
Some docs change requests, looks good overall.

mattclay commented:
Closing and re-opening to trigger CI.

gedl commented:
A couple of the provided examples specify HTTP as the protocol. However it seems the module is case sentive, producing the following error message:

```
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "msg": "value of protocol must be one of: http,https, got: HTTP"}
```

jsteinich commented:
The documented return values appear to be inside `target_group`, though the documentation makes it sound like they are top level elements.

wimnat committed:
Scope ec2 imports, add better doco, fix up examples, allow for both upper and lower case in protocol

wimnat commented:
@gedl - module should accept both case now but i've updated the examples anyway to always use lowercase.  Region is not specified in the doc but that is common across AWS modules because it isn't strictly required if you've configured it elsewhere such as in your boto3 profile or as an env variable.

@jsteinich - i have added the target_group attribute to better reflect what is returned.

Please comment with a 'shipit' if you think this module is good to go

wimnat commented:
@ryansb any idea on the CI error? :/

mattclay commented:
@wimnat There was a bug in the module validator that was masking the real issue. It should be fixed now, so I've restarted CI. When it finishes we should see the actual cause of the failure.

mattclay commented:
The test `ansible-test sanity --test validate-modules` failed with the following error:

```
lib/ansible/modules/cloud/amazon/elb_target_group.py:0:0: E302 DOCUMENTATION is not valid YAML. Line 83 column 134
```

wimnat committed:
Yaml fixes

wimnat committed:
ci fix

wimnat commented:
ready_for_review

dominics commented:
My main use case is to register/deregister targets from an ALB during pre_tasks and post_tasks. This is possible for ELBs using `aws_elb` like so:

```
ec2_elb:
  instance_id: "{{ ec2_id }}"
  state: absent
```

This removes the specified target, and doesn't touch the others.

But this module doesn't seem to allow you to safely register/deregister targets from the group. Passing a single target will remove any not mentioned. The best you can do is prepare a list of current targets yourself, and hope it doesn't change between when you retrieved it and when you pass it into the `targets` array.

All in all, this module currently allows idempotent management of the _group_ but not of the targets themselves. It seems to me that registering/deregistering specific targets is what a large proportion of the demand from https://github.com/ansible/ansible-modules-core/issues/4857 is asking for. Yet this doesn't seem to be dealt with yet? Is this intentional?

wimnat commented:
@dominics I can extend target group module or write a new one to just manage registered targets. What would be your preference?

I lean towards the latter to try and keep the modules simple

dominics commented:
@wimnat Yeah, a new one would make a lot of sense

Commit:
Added targets parameter to doc and added new modify_targets flag to prevent runs from modifying targets

wimnat committed:
Update to metadata_version

wimnat committed:
Update to metadata support

FlorinAndrei commented:
Managing (register / deregister) one instance at a time in a target group is important if you're migrating from ELBs to ALBs. It's crucial to have that capability if your current rolling deployment playbooks with ELBs have already been doing that for a while.

gundalow commented on lib/ansible/modules/cloud/amazon/elb_target_group.py:
```diff
@@ -0,0 +1,527 @@
+description:
+    - "Manage an AWS Application Elastic Load Balancer target group. See \
+    U(http://docs.aws.amazon.com/elasticloadbalancing/latest/application/load-balancer-target-groups.html) for details."
+version_added: "2.3"
```
`"2.4"`

gundalow commented on lib/ansible/modules/cloud/amazon/elb_target_group.py:
```diff
@@ -0,0 +1,527 @@
+    description:
+      - "The port on which the targets receive traffic. This port is used unless you specify a port override when registering the target. Required if \
+      state=present."
+    required: false
```
`required: false` lines can be removed as they are the default.

s-hertel commented on lib/ansible/modules/cloud/amazon/elb_target_group.py:
```diff
@@ -0,0 +1,527 @@
+        if e.response['Error']['Code'] == 'TargetGroupNotFound':
+            return None
+        else:
+            module.fail_json(msg=e.message, **camel_dict_to_snake_dict(e.response))
```
Can traceback be added to the exception handling in this module? i.e. `module.fail_json(msg=e.message, exception=traceback.format_exc(), **camel_dict_to_snake_dict(e.response))`

s-hertel replied:
There are a number of places this can be added.

wimnat replied:
Done

s-hertel commented on lib/ansible/modules/cloud/amazon/elb_target_group.py:
```diff
@@ -0,0 +1,527 @@
+    params['VpcId'] = module.params.get("vpc_id")
+
+    # If health check path not None, set health check attributes
+    if module.params.get("health_check_path") is not None:
```
Instead of `not None:` this can just be `if module.params.get("health_check_path"):`? This is done in a lot of other places as well.

wimnat replied:
TBH i prefer the slightly more verbose is not None as it handles cases such as empty lists.  Although these parameters may not allow for empty lists here, it is just my coding style.

s-hertel commented on lib/ansible/modules/cloud/amazon/elb_target_group.py:
```diff
@@ -0,0 +1,527 @@
+        if module.params.get("health_check_port") is not None:
+            params['HealthCheckPort'] = str(module.params.get("health_check_port"))
+
+        if module.params.get("health_check_interval") is not None:
```
Will this ever be None? It defaults to 30, and the minimum allowed by AWS is 5. I think we don't need to check 'is not None'.

wimnat replied:
I've removed the defaults because they are not a good idea - if someone didn't specify the parameter but changed it via the AWS console then the module would end up changing it.

s-hertel commented on lib/ansible/modules/cloud/amazon/elb_target_group.py:
```diff
@@ -0,0 +1,527 @@
+            health_check_port=dict(required=False, type='int'),
+            health_check_path=dict(required=False, default=None, type='str'),
+            health_check_interval=dict(required=False, default=30, type='int'),
+            health_check_timeout=dict(required=False, default=5, type='int'),
```
Maybe there should maybe be a check somewhere that health_check_interval > health_check_timeout? If the user sets health_check_interval to 5 (the minimum allowed by AWS) it will raise an ugly error. Or perhaps just documented.

wimnat replied:
AWS will return a suitable message and throw an exception if this is the case

s-hertel replied:
Yes. It's ugly though. No changes requested here I guess, just an observation.

s-hertel commented on lib/ansible/modules/cloud/amazon/elb_target_group.py:
```diff
@@ -0,0 +1,527 @@
+        if module.params.get("health_check_interval") is not None:
+            params['HealthCheckIntervalSeconds'] = module.params.get("health_check_interval")
+
+        if module.params.get("health_check_timeout") is not None:
```
Same as above.

s-hertel commented on lib/ansible/modules/cloud/amazon/elb_target_group.py:
```diff
@@ -0,0 +1,527 @@
+        if module.params.get("health_check_timeout") is not None:
+            params['HealthCheckTimeoutSeconds'] = module.params.get("health_check_timeout")
+
+        if module.params.get("healthy_threshold_count") is not None:
```
Same as the above.

s-hertel commented on lib/ansible/modules/cloud/amazon/elb_target_group.py:
```diff
@@ -0,0 +1,527 @@
+        if module.params.get("healthy_threshold_count") is not None:
+            params['HealthyThresholdCount'] = module.params.get("healthy_threshold_count")
+
+        if module.params.get("unhealthy_threshold_count") is not None:
```
Same as the above.

s-hertel commented on lib/ansible/modules/cloud/amazon/elb_target_group.py:
```diff
@@ -0,0 +1,527 @@
+        if module.params.get("unhealthy_threshold_count") is not None:
+            params['UnhealthyThresholdCount'] = module.params.get("unhealthy_threshold_count")
+
+        if module.params.get("successful_response_codes") is not None:
```
If this is ever None then line [341](https://github.com/ansible/ansible/pull/19492/files#diff-c008972f0ccccd4a4d9b8c1de95c4d08R341) is going to raise an ugly error. Since this defaults to 200 I think you should remove the check here, since if they set it to an empty string it will cause problems anyway (and they should probably know better).

wimnat commented:
@s-hertel I'm on holiday at the moment but back in 10 days. I'll go through your comments then. Thanks

s-hertel commented:
@wimnat Okay. Enjoy your holiday. :)

s-hertel commented:
@wimnat If you don't have time to work on this do you mind if I do? I'd love to see this get merged.

wimnat committed:
Remove defaults for some params. Add tags parameter

wimnat commented:
@s-hertel i think we're almost there on this one.  Just need that helper PR merged so I can do the tags here and then I think it's pretty much done.

wimnat committed:
Use paginator to get target groups

wimnat committed:
Add tag support

wimnat commented:
ok tags now done.

Final reviews welcome.

Module `elb_target` for just adding or removing a target coming soon

jbscare commented:
I tried this out just now; thanks so much for writing it!

> Module elb_target for just adding or removing a target coming soon

That'd be great.

One other thing that this module doesn't do: Set "target group attributes", the things that https://docs.aws.amazon.com/cli/latest/reference/elbv2/modify-target-group-attributes.html lets you modify. Any chance you'd be up for adding that functionality here? In the interest of getting this merged, I'd be happy to make a separate issue for that if you'd prefer.

willthames commented on lib/ansible/modules/cloud/amazon/elb_target_group.py:
```diff
@@ -0,0 +1,595 @@
+    choices: [ 'http', 'https' ]
+  port:
+    description:
+      - "The port on which the targets receive traffic. This port is used unless you specify a port override when registering the target. Required if \
```
YAML allows simpler line breaking than using `\`

```
  description:
  - The port on which the targets receive traffic. This port is used unless you specify a port 
    override when registering the target. Required if I(state) is C(present)
```

willthames commented on lib/ansible/modules/cloud/amazon/elb_target_group.py:
```diff
@@ -0,0 +1,595 @@
+    required: true
+  protocol:
+    description:
+      - The protocol to use for routing traffic to the targets. Required if state=present.
```
I(state) is C(present)

willthames commented on lib/ansible/modules/cloud/amazon/elb_target_group.py:
```diff
@@ -0,0 +1,595 @@
+module: elb_target_group
+short_description: Manage a target group for an Application load balancer
+description:
+    - "Manage an AWS Application Elastic Load Balancer target group. See \
```
Use simpler line wrapping (see comment below)

wimnat committed:
Add tg attributes to module

wimnat commented:
@jbscare tg attributes added

wimnat commented:
@willthames comments addressed :)

Commit:
Quote multilines

wimnat commented:
bot_broken - the spaces are required otherwise the docs will not have spaces.

s-hertel commented:
shipit

gundalow commented:
Regarding the trailing whitespace issues I think of you:
1) Remove the double quotes 
2) Remove the trailing whitespace
That should fix the issue

willthames commented on lib/ansible/modules/cloud/amazon/elb_target_group.py:
```diff
@@ -0,0 +1,682 @@
+            stickiness_type=dict(default='lb_cookie', type='str'),
+            stickiness_lb_cookie_duration=dict(type='int'),
+            state=dict(required=True, choices=['present', 'absent'], type='str'),
+            successful_response_codes=dict(required=False, type='str'),
```
Can we make this a `list`. "200,250-299" should still work then, but so should

```
- 200
- "250-299"

willthames commented on lib/ansible/modules/cloud/amazon/elb_target_group.py:
```diff
@@ -0,0 +1,682 @@
+            health_check_protocol=dict(required=False, choices=['http', 'https', 'HTTP', 'HTTPS'], type='str'),
+            health_check_port=dict(required=False, type='int'),
+            health_check_path=dict(required=False, default=None, type='str'),
+            health_check_interval=dict(required=False, type='int'),
```
let's get rid of unnecessary `required=False`s

wimnat committed:
Remove unnecessary defaults. Fix multiline

willthames commented on lib/ansible/modules/cloud/amazon/elb_target_group.py:
```diff
@@ -0,0 +1,682 @@
+module: elb_target_group
+short_description: Manage a target group for an Application load balancer
+description:
+    - "Manage an AWS Application Elastic Load Balancer target group. See 
```
One remaining whitespace fix here - shippable is still complaining:

```
2017-05-28 08:06:52 ERROR: Found 1 pep8 issue(s) which need to be resolved:
2017-05-28 08:06:52 ERROR: lib/ansible/modules/cloud/amazon/elb_target_group.py:26:73: W291 trailing whitespace (100%)
```

Commit:
Fix line ending

wimnat commented:
!bot_broken

ryansb commented:
shipit
